File: tests/fixtures/sample-project/src/services/user-service.ts
/**
 * User Service
 */

import { User, UserID, CreateUserData, UpdateUserData, UserRole } from '../types/user';
import { ApiResponse, ApiError } from '../types/api';
import { ValidationResult } from '../types/common';
import { UserModel } from '../models/user-model';

export class UserService {
  private users: Map<UserID, UserModel> = new Map();

  async createUser(data: CreateUserData): Promise<ApiResponse<User>> {
    try {
      // 建立 User 物件
      const user: User = {
        ...data,
        id: this.generateId(),
        createdAt: new Date(),
        updatedAt: new Date()
      };

      // 驗證
      const userModel = new UserModel(user);
      const validation = userModel.validate();

      if (!validation.isValid) {
        return this.errorResponse('VALIDATION_ERROR', 'User validation failed', validation);
      }

      // 檢查 email 是否已存在
      if (this.findByEmail(user.email)) {
        return this.errorResponse('EMAIL_EXISTS', 'Email already exists');
      }

      // 儲存
      this.users.set(user.id, userModel);

      return this.successResponse(user);
    } catch (error) {
      return this.errorResponse('CREATE_FAILED', 'Failed to create user', error);
    }
  }

  async getUser(userId: UserID): Promise<ApiResponse<User>> {
    const userModel = this.users.get(userId);
    if (!userModel) {
      return this.errorResponse('USER_NOT_FOUND', `User ${userId} not found`);
    }

    return this.successResponse(userModel.getData());
  }

  async updateUser(userId: UserID, updates: UpdateUserData): Promise<ApiResponse<User>> {
    const userModel = this.users.get(userId);
    if (!userModel) {
      return this.errorResponse('USER_NOT_FOUND', `User ${userId} not found`);
    }

    userModel.update({ ...updates, updatedAt: new Date() });

    const validation = userModel.validate();
    if (!validation.isValid) {
      return this.errorResponse('VALIDATION_ERROR', 'User validation failed', validation);
    }

    return this.successResponse(userModel.getData());
  }

  async deleteUser(userId: UserID): Promise<ApiResponse<boolean>> {
    if (!this.users.has(userId)) {
      return this.errorResponse('USER_NOT_FOUND', `User ${userId} not found`);
    }

    this.users.delete(userId);
    return this.successResponse(true);
  }

  async listUsers(): Promise<ApiResponse<User[]>> {
    const users = Array.from(this.users.values()).map(model => model.getData());
    return this.successResponse(users);
  }

  findByEmail(email: string): User | undefined {
    for (const userModel of this.users.values()) {
      const user = userModel.getData();
      if (user.email === email) {
        return user;
      }
    }
    return undefined;
  }

  findByRole(role: UserRole): User[] {
    const result: User[] = [];
    for (const userModel of this.users.values()) {
      const user = userModel.getData();
      if (user.role === role) {
        result.push(user);
      }
    }
    return result;
  }

  private generateId(): string {
    return `user_${Date.now()}_${Math.random().toString(36).substring(7)}`;
  }

  private successResponse<T>(data: T): ApiResponse<T> {
    return {
      success: true,
      data,
      timestamp: Date.now()
    };
  }

  private errorResponse(code: string, message: string, details?: unknown): ApiResponse<never> {
    const error: ApiError = {
      code,
      message,
      details: details as Record<string, unknown>
    };

    return {
      success: false,
      data: undefined as never,
      error,
      timestamp: Date.now()
    };
  }
}

File: tests/fixtures/swift-sample-project/Sources/SwiftSampleApp/Services/UserService.swift
import Foundation

/// User service errors
enum UserServiceError: Error {
    case userNotFound
    case invalidEmail
    case duplicateEmail
    case networkError(Error)
}

/// User service protocol
protocol UserServiceProtocol {
    func getUser(id: String) async throws -> User
    func getAllUsers() async throws -> [User]
    func createUser(name: String, email: String, role: UserRole) async throws -> User
    func updateUser(_ user: User) async throws -> User
    func deleteUser(id: String) async throws -> Bool
}

/// User service implementation
final class UserService: UserServiceProtocol {
    /// Simulated users storage
    private var users: [String: User] = [:]

    /// Get user by ID
    func getUser(id: String) async throws -> User {
        guard let user = users[id] else {
            throw UserServiceError.userNotFound
        }
        return user
    }

    /// Get all users
    func getAllUsers() async throws -> [User] {
        Array(users.values)
    }

    /// Create new user
    func createUser(name: String, email: String, role: UserRole) async throws -> User {
        guard isValidEmail(email) else {
            throw UserServiceError.invalidEmail
        }

        let user = User(
            id: UUID().uuidString,
            name: name,
            email: email,
            role: role,
            createdAt: Date()
        )
        users[user.id] = user
        return user
    }

    /// Update user
    func updateUser(_ user: User) async throws -> User {
        guard users[user.id] != nil else {
            throw UserServiceError.userNotFound
        }
        users[user.id] = user
        return user
    }

    /// Delete user
    func deleteUser(id: String) async throws -> Bool {
        guard users[id] != nil else {
            throw UserServiceError.userNotFound
        }
        users.removeValue(forKey: id)
        return true
    }

    /// Validate email format
    private func isValidEmail(_ email: String) -> Bool {
        let pattern = "[A-Z0-9a-z._%+-]+@[A-Za-z0-9.-]+\\.[A-Za-z]{2,64}"
        return email.range(of: pattern, options: .regularExpression) != nil
    }
}
